webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala

```diff
@@ -771,61 +771,69 @@ class ComplianceAPIService(
-      t2        <- currentTimeMillis
-      _         <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - nodeGroupRepo.getAllNodeIds in ${t2 - t1} ms")
-
+      t1           <- currentTimeMillis
       directiveLib <- directiveRepo.getFullDirectiveLibrary()
-      t3           <- currentTimeMillis
-      _            <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - getFullDirectiveLibrary in ${t3 - t2} ms")
+      t2           <- currentTimeMillis
+      _            <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - getFullDirectiveLibrary in ${t2 - t1} ms")
 
       nodeFacts <- nodeFactRepos.getAll()
-      t4        <- currentTimeMillis
-      _         <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - nodeFactRepo.getAll() in ${t4 - t3} ms")
+      t3        <- currentTimeMillis
+      _         <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - nodeFactRepo.getAll() in ${t3 - t2} ms")
 
       compliance <- getGlobalComplianceMode
-      t5         <- currentTimeMillis
-      _          <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - getGlobalComplianceMode in ${t5 - t4} ms")
+      t4         <- currentTimeMillis
+      _          <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - getGlobalComplianceMode in ${t4 - t3} ms")
 
       rules <- rulesRepo.getAll()
-      t6    <- currentTimeMillis
-      _     <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - getAllRules in ${t6 - t5} ms")
+      t5    <- currentTimeMillis
+      _     <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - getAllRules in ${t5 - t4} ms")
 
       groupLib <- nodeGroupRepo.getFullGroupLibrary()
-      t7       <- currentTimeMillis
-      _        <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - getFullGroupLibrary in ${t7 - t6} ms")
+      t6       <- currentTimeMillis
+      _        <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - getFullGroupLibrary in ${t6 - t5} ms")
 
-      // - global compliance : does the group contain a node that any rule applied to ? yes: include the rule in the compliance, and only nodes that belong to group
-      // - targeted compliance : does the rule target the group directly (for group targets : includes it, and not in exclude) ? yes: same
+      globalMode <- getGlobalPolicyMode
 
-      // filter our rules that are applicable to any node in this group and get their compliances
-      currentGroupNodeIds = allGroups.getOrElse(nodeGroup.id, Chunk.empty)
+      nodeSettings = nodeFacts.mapValues(_.rudderSettings).toMap
 
-      targetedNodeIdsByRuleId = {
+      // A map for constant time access for the set of targeted nodes and policy mode:  a Map[RuleId, (Chunk[NodeId], Option[PolicyMode])]
+      // The set is reused to directly compute the policy mode of the rule
+      allRuleInfo = {
+        val allGroups = groupLib.allGroups.map { case (k, v) => (k, Chunk.from(v.nodeGroup.serverList)) }
```

Comment:
you certainly ought to time this part

Reply:
I'm not sure this would be interesting for this specific line of code. This operation is included with the 

```
_ <- TimingDebugLoggerPure.trace(s"getByNodeGroupCompliance - filterGlobalOrTargetedRules in ${t7 - t6} ms")
```
line for logging, which includes the timing for "all computations needed to filter the rules for the group", so this one included

Reply:
Do you mean to separate the timing in two different parts : "getRuleInfos" (policy mode and targeted nodes) and "filterRules" ?